### DISCUSSION TITLE
Help with a design with coupled motors

### DISCUSSION THREAD

**Discussion by cbullo** [google-deepmind/mujoco#2297] [Asking for Help] [answered]
### Intro

I'm playing with an idea for a robot.

### My setup

Playing with CoppeliaSim

### My question

![sketch](https://github.com/user-attachments/assets/d8a5504c-ef99-4855-b61a-df47c822284a)

I would like to simulate the mechanism drafted in the image.
- Solid line is a timing belt
- Red arrows mark two motors
- Other circles are pulleys
- Green line is an axis of a prismatic joint

The idea is that if motors M1 and M2 turn with the same velocity, then the position of a prismatic joint doesn't change. And the angular velocity of top and bottom pulleys is an average of the two motors.

Somewhat inaccurate description is then:
$$Angle_{bpulley}=\frac{Angle_{M1} + Angle_{M2}}{2}$$
$$Translation_{rod}=Ratio * (Angle_{M1}-Angle_{M2})$$

This means that both the pulleys rotation and position of prismatic joint depend on both motors simultaneously.

I can't find a way to express this dependencies in MuJoCo, as it only supports a single joint dependency constraints.

Do you have any suggestions? 

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment, marked as answer]
You can do this with an equality constraint on a fixed tendon.

**Reply by cbullo** [reply]
Hmm, I'm not sure if I can. What should be the constraint length in this case? Also how to simulate the top/bottom pulleys rotation? They are coupled with both motors.

**Reply by yuvaltassa** [reply]
I think these are 3 separate fixed tendons:
1. Sum of linear joints should be fixed.
2. Top pulley rotation = average of motors.
3. Bottom pulley rotation = average of motors.

I think 🙂

**Reply by cbullo** [reply]
Ah, I understood what you meant now:

> fixed/⁠joint (*)
> This element adds a joint to the computation of the fixed tendon length. The position or angle of each included joint is multiplied by the corresponding coef value, and added up to obtain the tendon length.

I didn't realize there can be more than one joint referenced by a fixed tendon.

I'll give it a try!

**Reply by cbullo** [reply]
It worked, here's a sketch of my robot. Also, car model from examples was a good starting point.

```
<mujoco>
    <compiler autolimits="true" />

    <asset>
        <texture name="grid" type="2d" builtin="checker" width="512" height="512" rgb1=".1 .2 .3"
            rgb2=".2 .3 .4" />
        <material name="grid" texture="grid" texrepeat="1 1" texuniform="true" reflectance=".2" />
    </asset>

    <default>
        <joint damping=".03" actuatorfrcrange="-0.5 0.5" />
        <default class="wheel">
            <geom type="cylinder" size=".033 .014" rgba=".5 .5 1 1" />
        </default>
        <default class="rod">
            <geom type="box" size=".01 .01 .17" rgba=".5 .5 1 1" density="100"/>
        </default>
    </default>

    <worldbody>
        <geom type="plane" size="3 3 .01" material="grid" />
        <body name="robot" pos="0 0 .203" >
            <freejoint />
            <geom type="box" size=".11 .08 .07" mass=".1"/>
            <body name="left_rod" pos="0 -.04 0">
                <joint name="left_d" type="slide" range = "-0.065 0.065"/>
                <geom class="rod" />

                <body name="left wheel" pos="0 -0.07 -0.17" zaxis="0 1 0">
                    <joint name="left_r" />
                    <geom class="wheel" />
                </body>
            </body>
            <body name="right_rod" pos="0 .04 0">
                <joint name="right_d" type="slide" range = "-0.065 0.065"/>
                <geom class="rod" />

                <body name="right wheel" pos="0 0.07 -0.17" zaxis="0 1 0">
                    <joint name="right_r" />
                    <geom class="wheel" />
                </body>
            </body>
        </body>
    </worldbody>

    <tendon>
        <fixed name="left_m1">
            <joint joint="left_r" coef="1" />
            <joint joint="left_d" coef="1" />
        </fixed>
        <fixed name="left_m2">
            <joint joint="left_r" coef="1" />
            <joint joint="left_d" coef="-1" />
        </fixed>
        <fixed name="right_m1">
            <joint joint="right_r" coef="1" />
            <joint joint="right_d" coef="1" />
        </fixed>
        <fixed name="right_m2">
            <joint joint="right_r" coef="1" />
            <joint joint="right_d" coef="-1" />
        </fixed>
    </tendon>

    <actuator>
        <motor name="left_m1_motor" tendon="left_m1" ctrlrange="-100 100"/>
        <motor name="left_m2_motor" tendon="left_m2"  ctrlrange="-100 100"/>
        <motor name="right_m1_motor" tendon="right_m1" ctrlrange="-100 100"/>
        <motor name="right_m2_motor" tendon="right_m2" ctrlrange="-100 100"/>
    </actuator>

</mujoco>
```

**Reply by yuvaltassa** [reply]
That's great, though I'm surprised you didn't put equality constraints on some tendons?

Usually in this situation we find ourselves having two kinds of tendons: those whose length is constrained by equalities and those that are actuated, to split the forces/torques among the joints.

Good luck!

**Reply by cbullo** [reply]
I believe I could do it, because the rods themselves create length constrains. And the upper pulley is not really important for the simulation, so I removed it. 
In addition, I kind of 'inversed' the problem - fixed tendons are used to store total rotation of motors and coefficients of joints inside guarantee the correctness of 'inverted' joint relations, which luckily is linear. I hope that makes sense.

Don't know if this is best for correctness of dynamics, but it's good enough for me to play with.

Thanks for your help, again.
